### PR TITLE
[sanitizer] Extract SANITIZER_FREEBSD version of ThreadDescriptorSizeFallback

### PR DESCRIPTION
**Pull request by vitalybuka**
This should fix SANITIZER_FREEBSD and simplify
SANITIZER_GLIBC version.

Also the PR make readers aware of problematic
`ThreadDescriptorSizeFallback` for SANITIZER_FREEBSD.
Maybe it will encourage FreeBSD maintainers to
improve the functions, or prove that it's not needed at
all.

### CODE REVIEW

**Comment by vitalybuka** [comment]
`ThreadDescriptorSize()` Is used in a few FREEBSD branches, but it looks very wrong.

File: compiler-rt/lib/sanitizer_common/sanitizer_linux_libcdep.cpp [hunk 1]
```diff
@@ -304,6 +293,48 @@ static uptr ThreadDescriptorSizeFallback() {
...
+
+#    if defined(__aarch64__)
+  // The sizeof (struct pthread) is the same from GLIBC 2.17 to 2.22.
+  return 1776;
+#    endif
+
+#    if defined(__powerpc64__)
+  return 1776;  // from glibc.ppc64le 2.20-8.fc21
+#    endif
+
```
Comment (rorth) [inline compiler-rt/lib/sanitizer_common/sanitizer_linux_libcdep.cpp:327]: I have a hard time believing that that's any point in this: apart from being glibc-specific, as of FreeBSD 14.1 there isn't a s390, sparc, mips, or loongarch64 port of FreeBSD.

**Comment by rorth** [comment]
FWIW, this does indeed fix the build failure on FreeBSD 14.0, without affecting Solaris.  Unfortunately, the build fails later for different reasons:
```
FAILED: projects/compiler-rt/lib/builtins/CMakeFiles/clang_rt.builtins-i386.dir/extendhfsf2.c.o
[...]
In file included from compiler-rt/lib/builtins/extendhfsf2.c:11:
In file included from compiler-rt/lib/builtins/fp_extend_impl.inc:38:
compiler-rt/lib/builtins/fp_extend.h:57:9: error
: _Float16 is not supported on this target
   57 | typedef _Float16 src_t;
      |         ^
1 error generated.
```
While `_Float16` is supported for 64-bit x86, it isn't for 32-bit (since that doesn't support SSE2 for whatever reason).  However, the `builtins` `CMakeFile.list` only checks for `_Float16` support for the default multilib, missing this issue.

File: compiler-rt/lib/sanitizer_common/sanitizer_linux_libcdep.cpp (hunk 1 above)
Comment (vitalybuka) [reply compiler-rt/lib/sanitizer_common/sanitizer_linux_libcdep.cpp:327]: If you are confident, I can drop them in followup PR, so bisecting could be easiers.

According to https://www.freebsd.org/platforms/
Recent Tier 1 are only x86_64 and aarch64, the former works with 0.

I guess implication of having these values invalid is incorrect DTLS detection which will false positive for msan and lsan.

**Comment by vitalybuka** [comment]
> FWIW, this does indeed fix the build failure on FreeBSD 14.0, without affecting Solaris. Unfortunately, the build fails later for different reasons:
> 
> ```
> FAILED: projects/compiler-rt/lib/builtins/CMakeFiles/clang_rt.builtins-i386.dir/extendhfsf2.c.o
> [...]
> In file included from compiler-rt/lib/builtins/extendhfsf2.c:11:
> In file included from compiler-rt/lib/builtins/fp_extend_impl.inc:38:
> compiler-rt/lib/builtins/fp_extend.h:57:9: error
> : _Float16 is not supported on this target
>    57 | typedef _Float16 src_t;
>       |         ^
> 1 error generated.
> ```
> 
> While `_Float16` is supported for 64-bit x86, it isn't for 32-bit (since that doesn't support SSE2 for whatever reason). However, the `builtins` `CMakeFile.list` only checks for `_Float16` support for the default multilib, missing this issue.

_Float16 is unrelated to the PR issue?

**Review by fmayer — COMMENTED** [review]
Change mechanically LGTM, but I can't comment on the FreeBSD

**Pull request merged**